switch case without default no longer compiles in visual studio 2019 [c#]

VSF_TYPE_MARKDOWNThe following code compiles fine in Visual Studio 2017, but not in Visual Studio 2019. Is there a default compiler option that has changed which can resolve this?

    using System;
    
    namespace NoDefaultCase
    {
        public static class Program
        {
            public const string Database = &quot;MongoDB&quot;;
    
            public static string GetDb()
            {
                switch (Database)
                {
                    case &quot;MongoDB&quot;:
                        return Database;
                }
            }
    
            public static void Main(string[] args)
            {
                Console.WriteLine($&quot;The value returned is \&quot;{GetDb()}\&quot;.&quot;);
                Console.WriteLine(&quot;Press a key to exit&quot;);
                Console.ReadKey();
            }
        }
    }

There is also a Stack Overflow ticket for this item:
https://stackoverflow.com/questions/55541269/why-does-a-switch-case-statement-on-a-string-constant-require-a-default-in-visua

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/520522/switch-case-without-default-no-longer-compiles-in.html
VSTS ticketId: 841662_
_These are the original issue comments:_

Martin Eyles on 4/6/2019, 05:37 AM (10 days ago): <p>In Visual Studio 2017, the output of the build is:<br><br>    1&gt;------ Rebuild All started: Project: NoDefaultCase, Configuration: Debug Any CPU ------<br>    1&gt;  NoDefaultCase -&gt; C:\Users\MartinEyles\source\repos\NoDefaultCase\NoDefaultCase\bin\Debug\NoDefaultCase.exe<br>    ========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========<br><br>In Visual Studio 2019, the output of the build is:<br><br>    1&gt;------ Rebuild All started: Project: NoDefaultCase, Configuration: Debug Any CPU ------<br>    1&gt;C:\Users\MartinEyles\source\repos\NoDefaultCase\NoDefaultCase\Program.cs(9,30,9,35): error CS0161: 'Program.GetDb()': not all code paths return a value<br>    ========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========</p>

Visual Studio Feedback System on 4/8/2019, 02:55 AM (8 days ago): <p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

Martin Eyles on 4/8/2019, 03:07 AM (8 days ago): <p>A github repository containing the example solution can be found at <a target='_blank' href="https://github.com/martineyles/NoDefaultCase">https://github.com/martineyles/NoDefaultCase</a> This includes an archive of the example solution in the state before it was added to github (NoDefaultCase - Before Github.zip).</p>

Martin Eyles on 4/8/2019, 05:59 AM (8 days ago): <p>I experience this issue in </p><p>Microsoft Visual Studio Enterprise 2019 <br>Version 16.0.0<br>VisualStudio.16.Release/16.0.0+28729.10<br>Microsoft .NET Framework Version 4.7.03056'</p><p>It has been suggested by others in the Stack Overflow community that a preview version of Visual Studio 2019 behaves the same way as Visual Studio 2017</p>

Martin Eyles on 4/12/2019, 05:24 AM (4 days ago): <p>The issue is still present in:<br><br>Microsoft Visual Studio Enterprise 2019<br>Version 16.0.1<br>VisualStudio.16.Release/16.0.1+28803.156<br>Microsoft .NET Framework Version 4.7.03056</p>

Martin Eyles on 4/12/2019, 05:26 AM (4 days ago): <p>Stack overflow users has suggested that this does not meet the C#  specification, and is likely to be either a bug or a change in the  specification that hasn't yet been documented.</p>

Martin Eyles on 4/12/2019, 05:29 AM (4 days ago): <p>To quote a user with very good reputation on stack overflow:</p><p>The <a target='_blank' href="https://www.ecma-international.org/publications/standards/Ecma-334.htm">ECMA C# 5 standard</a> section 13.8.3 describes the reachability of the end of a switch statement:</p><p>  
</p><blockquote>   
<p>The end point of a switch statement is reachable if at least one of the following is true:</p><p>      
</p><ul><li>The switch statement contains a reachable break statement that exits the switch statement.</li><li>The switch statement is reachable, the switch expression is a non-constant value, and no default label is present.</li><li>The switch statement is reachable, the switch expression is a  constant value that doesn’t match any case label, and no default label  is present.</li></ul></blockquote><p>  
</p><p>None of these seem to be the case in your example:</p><p>  
</p><ul><li>There are no break statements</li><li>The switch expression is a constant value</li><li>The constant value <em>does</em> match a case label</li></ul><p>  
</p><p>So with C# 5 rules, the end point of this switch statement is <em>not</em> reachable, and it should compile with no problems. The <a target='_blank' href="https://github.com/dotnet/csharplang/blob/master/spec/statements.md">draft specification in GitHub</a> has the same text, so it doesn't look like it's changed there yet...</p>

_These are the original issue solutions:_
(no solutions)

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs
@@ -63,7 +63,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 Join(ref afterSwitchState, ref this.State);
             }
 
-            if (reachableLabels.Contains(node.BreakLabel) || node.DefaultLabel == null && IsTraditionalSwitch(node))
+            if (node.DecisionDag.ReachableLabels.Contains(node.BreakLabel) ||
+                (node.DefaultLabel == null && node.Expression.ConstantValue is null && IsTraditionalSwitch(node)))
             {
                 Join(ref afterSwitchState, ref initialState);
             }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs
@@ -63,7 +63,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 Join(ref afterSwitchState, ref this.State);
             }
 
-            if (node.DecisionDag.ReachableLabels.Contains(node.BreakLabel) ||
+            if (reachableLabels.Contains(node.BreakLabel) ||
                 (node.DefaultLabel == null && node.Expression.ConstantValue is null && IsTraditionalSwitch(node)))
             {
                 Join(ref afterSwitchState, ref initialState);

--- a/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
@@ -132,7 +132,7 @@ class C
         switch (S)
         {
             case ""def"":
-                return S;
+                return S; // 1
             default:
                 return S;
         }
@@ -143,7 +143,7 @@ class C
         switch (S)
         {
             case ""def"":
-                return S;
+                return S; // 2
         }
         // error
     }
@@ -152,24 +152,24 @@ class C
             var comp = CreateCompilation(src);
             comp.VerifyDiagnostics(
                 // (40,17): warning CS0162: Unreachable code detected
-                //                 return S;
+                //                 return S; // 1
                 Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(40, 17),
                 // (46,26): error CS0161: 'C.M5()': not all code paths return a value
                 //     public static string M5()
                 Diagnostic(ErrorCode.ERR_ReturnExpected, "M5").WithArguments("C.M5()").WithLocation(46, 26),
                 // (51,17): warning CS0162: Unreachable code detected
-                //                 return S;
+                //                 return S; // 2
                 Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(51, 17));
             comp = CreateCompilation(src, parseOptions: TestOptions.Regular7_3);
             comp.VerifyDiagnostics(
                 // (40,17): warning CS0162: Unreachable code detected
-                //                 return S;
+                //                 return S; // 1
                 Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(40, 17),
                 // (46,26): error CS0161: 'C.M5()': not all code paths return a value
                 //     public static string M5()
                 Diagnostic(ErrorCode.ERR_ReturnExpected, "M5").WithArguments("C.M5()").WithLocation(46, 26),
                 // (51,17): warning CS0162: Unreachable code detected
-                //                 return S;
+                //                 return S; // 2
                 Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(51, 17));
         }
 

--- a/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs
@@ -90,6 +90,90 @@ public class DATest : DATestBase {
 }";
 
         [Fact]
+        [WorkItem(35011, "https://github.com/dotnet/roslyn/issues/35011")]
+        public void SwitchConstantUnreachable()
+        {
+            var src = @"
+class C
+{
+    const string S = ""abc"";
+
+    public static string M1()
+    {
+        switch (S)
+        {
+            case ""abc"":
+                return S;
+        }
+    }
+
+    public static string M2()
+    {
+        const string S2 = S + """";
+        switch (S)
+        {
+            case S2:
+                return S;
+        }
+    }
+
+    public static string M3()
+    {
+        const int I = 11;
+        switch (I)
+        {
+            case 11:
+                return S;
+        }
+    }
+
+    public static string M4()
+    {
+        switch (S)
+        {
+            case ""def"":
+                return S;
+            default:
+                return S;
+        }
+    }
+
+    public static string M5()
+    {
+        switch (S)
+        {
+            case ""def"":
+                return S;
+        }
+        // error
+    }
+}";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (40,17): warning CS0162: Unreachable code detected
+                //                 return S;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(40, 17),
+                // (46,26): error CS0161: 'C.M5()': not all code paths return a value
+                //     public static string M5()
+                Diagnostic(ErrorCode.ERR_ReturnExpected, "M5").WithArguments("C.M5()").WithLocation(46, 26),
+                // (51,17): warning CS0162: Unreachable code detected
+                //                 return S;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(51, 17));
+            comp = CreateCompilation(src, parseOptions: TestOptions.Regular7_3);
+            comp.VerifyDiagnostics(
+                // (40,17): warning CS0162: Unreachable code detected
+                //                 return S;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(40, 17),
+                // (46,26): error CS0161: 'C.M5()': not all code paths return a value
+                //     public static string M5()
+                Diagnostic(ErrorCode.ERR_ReturnExpected, "M5").WithArguments("C.M5()").WithLocation(46, 26),
+                // (51,17): warning CS0162: Unreachable code detected
+                //                 return S;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(51, 17));
+        }
+
+        [Fact]
         public void General()
         {
             var source = prefix + @"

